packages/core/src/internal/runSaga.js

```diff
@@ -77,8 +85,7 @@ export function runSaga(options, saga, ...args) {
   }
 
   const env = {
-    stdChannel: channel,
-    dispatch: wrapSagaDispatch(dispatch),
+    channel: { ...channel, put: wrapSagaDispatch(dispatch || channel.put) },
```

Comment:
Swap `channel.put` here so that `runPutEffect` just need to call `channel.put` (no need to know the `dispatch` any more).

`dispatch || channel.put` means that we are making the `dispatch` **optional**, when uses do not supply a `dispatch`, users can still use put/take through the underlying `env.channel`.

Reply:
doesnt it mean that if somebody would use custom `channel` through the option then his/her action wouldnt get dispatched automatically?

Reply:
Let me describe it clearer:

`sagaMiddlewareFactory.options` has now 6 fields (namely `context, channel, sagaMonitor, logger, onError, effectMiddlewares`), and all of them is optional. When connect to a redux store, `sagaMiddlewareFactory` can get `getState` and `dispatch`, and then pass them to `runSaga` along with the above 6 fields.

In cases that users are using `sagaMiddlewareFactory`, `options.dispatch` in `runSaga` is always not null, `put` effects will always hit reducers.

`runSaga.options` has now 8 fields, 2 fields more than `sagaMiddlewareFactory.options`, namely `dispatch` and `getState`.

In cases that users are directly using `runSaga`, `options.dispatch` behaves like a put-handler. **In the past**, users must pass the `dispatch` option, otherwise calls to the `dispatch` will throw an exception in [`runPutEffect`](https://github.com/redux-saga/redux-saga/blob/master/packages/core/src/internal/proc.js#L456). **Now**, `dispatch || channel.put` means that we are making the `dispatch` in `runSaga.options` optional, when uses do not supply a `dispatch`, put/take still works, and users can use it for inter-saga communication.

Reply:
`dispatch` is used in saga for emitting actions outside of its runtime
`channel.put` is used for emitting **into** saga runtime

Having them separate makes a nice distinction, although might be a little bit confusing.

Current solution has other problem though, by making this:
```js
channel: { ...channel, put: wrapSagaDispatch(dispatch || channel.put) },
```

> In cases that users are using sagaMiddlewareFactory, options.dispatch in runSaga is always not null, put effects will always hit reducers.

I've pushed out a change to the test to showcase the problem, you can check out that simply removing `dispatch` is causing a problem there.

Reply:
> 😕 Looks the expected result is wrong to me. Can you double check that?

You are right, I've made a silly mistake there 😅 

I think there might be edge cases around current solution, but Im not saying it's a wrong direction - we only might need to tweak it a little bit. I need to give this a little bit more thought, gonna experiment today with how it works exactly now and what can and what cannot be done to see if we can follow that direction.